### ISSUE TITLE
Incorrect position of widgets

### ISSUE THREAD

**Issue by Skryabind** [ducksboard/gridster.js#147]
Hello!
In some cases gridster places widgets in incorrect cells.
Simple example: http://jsfiddle.net/Skryabind/rPVPe/
You can see that element with data-row="2" and data-col="2" actually places on (1,1). Element with data-row="3" and data-col="2" places on (2,2).
This problem happens in other cases, but I did not find what they have in common.

**Comment by dustmoo** [comment]
Confirmed.

The problem seems to do with the order of the <li> elements.  If you order them in the same order of your grid rows there doesn't appear to be a problem. I am not sure what is happening in the code, as it shouldn't worry about order of the elements.

**Comment by badtant** [comment]
also having a problem with this

**Comment by badtant** [comment]
it's because gridster is trying to move widgets up. in add_to_gridmap (which is used to register widgets) you take the widgets below the one you're registering and move them up if possible. so in the fiddle above "3 2" is moved to "2 2". after that you try to register "2 2" which is then accupied by the "3 2" widget.

**Comment by dustmoo** [comment]
I found this happening when working on my fork. There are other times when
the script is calling "move_widget_up" multiple times and causing an
overlap. @badtant has it right.

I mostly reworked this in my fork.

**Comment by antonkulaga** [comment]
I've just discovered that I've got the same problem ( https://github.com/ducksboard/gridster.js/issues/156 )

> it's because gridster is trying to move widgets up. in add_to_gridmap (which is used to register widgets) 

It is very weird. Why should it try to move it up at all? If the programmer explicitly tells the widget to be placed in specified cells he has some rationale to do so, even if there are some free cells that are higher than the widget.

> I mostly reworked this in my fork.

Why did not you merge the official one with yours if the official ones has such critical bugs? I see you have an access rights to do so.

**Comment by dustmoo** [comment]
> Why did not you merge the official one with yours if the official ones has such critical bugs? I see you have an access rights to do so.

Because my functionality has changed quite a bit and I have not tested a merged version yet. My project has diverged from this project substantially and unfortunately I do not have the time to put them together right now.

**Comment by antonkulaga** [comment]
Thank you for the response!
I moved to your fork, widgets started to behave a bit differently, but the bug with wrong positioning https://github.com/ducksboard/gridster.js/issues/156 is still there =(

**Comment by dustmoo** [comment]
Do you have an example I can look at to troubleshoot?

**Comment by antonkulaga** [comment]
Thanks you for the fast answer, to be honest I did not expect it to be so fast=))

I will update https://github.com/denigma/semantictasks repository tomorrow with your gridster version. so you will be able to check it.

But the code that positions the widgets is the same ( https://github.com/denigma/semantictasks/blob/master/view_models/widget_board.coffee) as well as how it looks like ( https://f.cloud.github.com/assets/842436/383775/568c2826-a626-11e2-8165-5e1d8b4e455c.png )

**Comment by dustmoo** [comment]
Great, I will have a look and see what I can find and fix. :)

**Comment by antonkulaga** [comment]
By the way, could you tell me, please, how can I switch off dragging for elements inside widget block? I have some text elements there that can be selected by users and such selection may fire unwanted drag of the whole widget.

**Comment by dustmoo** [comment]
If you add a the class "static" to the widget it will not be draggable. It can be shifted down however unless you set disable moving larger widgets down. See the sample in the repo.

**Comment by antonkulaga** [comment]
Dustin, I updated the repo ( https://github.com/denigma/semantictasks ). Now it contains your version of gridster (by the way I like how it looks, you did a great job with animations there). When you compile the repo (it is based coffeescript, so you have to execute cake build to compile) you will see that in index.html the widget with search input is lower than it should be (according to https://github.com/denigma/semantictasks/blob/master/view_models/widget_board.coffee ).

**Comment by dustmoo** [comment]
Great, I'll look at it ASAP.

**Comment by dustmoo** [comment]
My apologies for not getting to this. I have some items sucking all of my
time. I havn't forgotten you.

Thanks,

Dustin

**Comment by badtant** [comment]
we've added a sorting of widgets on row before presenting them to gridster. this fixes the problem but now the "right way".

**Comment by dustmoo** [comment]
Thanks for the update. :) I am working on this... slowly. I only have a small portion of time to devote to gridster.

**Comment by ktw** [comment]
Had the exact same problem, using stored values. (From db)
Fixed it using some jQuery before and after Gridster init:

```
        /* Hotfix - Part 1 */
        var items = $(".gridster ul li");
        items .detach();

        /* Gridster setup code */
        var gridster = $(".gridster ul").gridster({
                    // Init code here...
        }).data('gridster');

        /* Hotfix - Part 2 */
        $.each(items , function (i, e) {
            var item = $(this);
            var columns = parseInt(item.attr("data-sizex"));
            var rows = parseInt(item.attr("data-sizey"));
            var col = parseInt(item.attr("data-col"));
            var row = parseInt(item.attr("data-row"));
            gridster.add_widget(item, columns, rows, col, row);
        });
```

**Comment by weberkai** [comment]
First, thank you very much! Gridster is great!
When I serialize, and try to execute a loop with add_widget to restore them, a few widgets get swapped.
I think if the first widget is below the second, during addition the first moves up as the cells are free, and then the second moves down as the cells are occupied. Suggestion for analisys: Maybe the method gridster.disable() could disable "pull widgets up" also, and then solve this if the user execute before adding?

**Comment by MelatoninUK** [comment]
@ktw:

Thanks Kim, that hotfix approach put me on the right path.

There were still one or two occasions when the ordering was rendering back incorrectly, but I found that ordering the list of items by data-row before adding them to the grid helped.

something like this before calling add_widget(), to make sure the top most items are added first

``` JavaScript
items.sort(function(a, b) {
    var rowA = parseInt(a.attributes[0].value);
    var rowB = parseInt(b.attributes[0].value);
    return rowA - rowB;
});
```

**Comment by oosterholt** [comment]
Maybe the serialize function can be adjusted to sort the widgets by row. 

This way the serialized data is in the 'correct order' to be re-added to the grid (from DB or something).

**Comment by harshmsp** [comment]
Hello @dustmoo ,
I have 12*10 widget. I want to set one box fix at first row and first column, that can not be drag or change the position, so how can we do that ? Please suggest your answer.

**Comment by dustmoo** [comment]
Currently this only (kind of) works on the dustmoo branch by using the “static” class. Did you try that?

**Comment by harshmsp** [comment]
Yes @dustmoo , But with that if other object can move to that position then it works and  I want to fix that layout of define row and column that not draggable and also other objects can not be take that position.
Please give your suggestion.

**Comment by TVDKoni** [comment]
Hello all. Just to share the information. I faced the same issue and found the reason based on gridster initialization:
        $(".gridster ul").gridster({
            widget_margins: [tileMargin, tileMargin],
            widget_base_dimensions: [tileSize, tileSize],

Problem was that my variables tileMargin and tileSize were strings. After changing to floats, the grid was looking just fine:
        $(".gridster ul").gridster({
            widget_margins: [parseFloat(tileMargin), parseFloat(tileMargin)],
            widget_base_dimensions: [parseFloat(tileSize), parseFloat(tileSize)],

**Comment by cssherry** [comment]
I was running into the same problem with elements overlapping each other on gridster. While sorting widgets before adding them fixed the problem, it produced some strange behavior (empty spaces where there weren't any visuals) and made the loading take a bit longer due to the multiple sorts we were conducting. 

There was another fix to our problem since the bug was caused by cell occupation information being over-written in gridster.gridmap when fn.add_faux_cell is called. I fixed it by making the following change:

```
fn.add_faux_cell = function(row, col) {
        var coords = $({
                        left: this.baseX + ((col - 1) * this.min_widget_width),
                        top: this.baseY + (row -1) * this.min_widget_height,
                        width: this.min_widget_width,
                        height: this.min_widget_height,
                        col: col,
                        row: row,
                        original_col: col,
                        original_row: row
                    }).coords();

        if (!$.isArray(this.gridmap[col])) {
            this.gridmap[col] = [];
        }

-       this.gridmap[col][row] = false;
+       this.gridmap[col][row] = this.gridmap[col][row] || false;
        this.faux_grid.push(coords);

        return this;
    };

```

**Comment by HamzaShahidd** [comment]
i am getting this error after updating the gridster.js file in my project 
this.gridmap is occuring undefined which is causing problems while adding the widget 


![gridster](https://user-images.githubusercontent.com/39696772/40611025-234008b4-628e-11e8-8547-2e59f8e5965d.PNG)

Kindly comment below if anyone faced the same issue
